designs/builtins-design.md

```diff
@@ -0,0 +1,201 @@
+# Built-in Libraries Design
+
+## Overview
+We will implement a series of standard libraries and built-in functions for ChocoPy. We want to add a preliminary support for `float()` and `...` (Ellipsis), which includes initialization and value representation. 
+We'd also like to extend `print()` to allow arbitrary length arguments, and `int() / bool()` for type conversions.
+
+Moreover, we plan to support a simplified version of Python's `import` statements through AST transformations and some compiler refactoring (detailed below); then we plan to implement some of the built-in libraries & functions in ChocoPy itself, utilizing the function / class definitions already in the compiler, plus the `import`  statement support. This design gives us the opportunity of self-hosting some of the built-in library, and at the same time still allows us to write some part of it in TypeScript or WASM.
+
+## Changes to the Compiler
+These are the potential changes to the compiler that we plan to make in the coming week to support the first iteration of built-in libraries.
+
+#### Compilation Pipeline
+
+The current compiler can be roughly divided into 4 stages:
+1. Parsing (py -> ast)
+2. Type Checking (ast -> typed ast)
+3. Lowering (typed ast -> IR)
+4. Codegen (IR -> WASM)
+and the compilation process follows the diagram below.
+
+```mermaid
+flowchart LR
+    src[(ChocoPy Source)] --parse--> ast{{AST}}
+    ast --type check--> tast{{Typed AST}} & tenv[Global Env]
+    tast & tenv --lowering--> ir([IR])
+    ir & tenv --codegen--> WASM
+```
+To implement our built-in library and simple `import` support, we propose to change the compiler to behave like the following new diagram, adding an **Import Handling** stage to perform code rewrite, AST transformation / re-parsing. 
+```mermaid
+flowchart LR
+    src[(ChocoPy Source)] --parse--> ast{{AST}}
+    ast & src ==import handling==> xast{{Expanded AST}}
+    xast --type check--> tast{{Typed AST}} & tenv[Global Env]
+    tast & tenv --lowering--> ir([IR])
+    ir & tenv --codegen--> WASM
+    style xast fill:#82CFFD
+```
+We will need to change the `run()` function in `runner.ts` and the `compile()` function in `compiler.ts`, but our updated compiler should be a drop-in replacement of the original one, and shouldn't affect any other team.
+
+#### Import Handling
+We plan to place the import handling stage right after parsing before doing the type checking. We will use the information in AST, such as `imports` and function calls, to preprocess the input source code. For example, if we get the following source in its AST format, we will be able to recognize the import statement and the corresponding function calls depending on that import:
+
+```python
+from math import gcd, lcm
+print(gcd(12, 16), lcm(12, 16))
+```
+
+Then we rewrite the AST to include our library implementation, essentially replacing the original source file with the updated one below:
```

Comment:
This is clever, but is a bad idea for a few reasons.

1. With transitive imports (a imports b imports c), what if `b` and `c` both import `d`? Won't that make two copies of `d`? What if `d` prints at the toplevel?
2. What if one of the imported functions relies on another function or global variable in the module as a helper, but isn't explicitly imported?
3. Consider duplicate classes with the above issue – will there be extra vtables created?
4. In the REPL, this will make each REPL entry be very long!
5. What if two _different_ modules define the same function and are imported into the same module?

(In C/C++, this is the reason for a whole bunch of nonsense with `#IFDEF`, etc.)

The strategy presented in the starter code, with `memory.wat` compiling to `memory.wasm` and then _linked_ through JavaScript might be a better one.

Then you could somehow communicate the updated _environment_ at the time of import (so from the compiler's point of view, the main thing that happens is a bigger starting environment and inserting some new WASM `import` statements).